### PR TITLE
feat: use Uint32Array for shuffling/committees

### PR DESCRIPTION
**Pull request by wemeetagain**
**Motivation**

Looking at heap snapshots of a running lodestar node, the memory usage of shufflings stood out as larger than necessary.

**Description**

Replace `number[]` with `Uint32Array` for `activeIndices`, `shuffling`, and `committees` in the `EpochShuffling`.
This has two benefits: tighter packing of the data (TypedArray vs Object) AND data sharing between `shuffling` and `committees`.
The tradeoff is that only indices < 2 ^ 32 (4.2 billion) can be represented here, rather than indices < 2 ^ 53. In practice this is plenty and not an issue.

### CODE REVIEW

File: packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts [hunk 1]
```diff
@@ -398,7 +398,7 @@ export function getNotSeenValidatorsFn(state: CachedBeaconStateAllForks): GetNot
       state
     );
 
-    return (epoch: Epoch, committee: number[]) => {
+    return (epoch: Epoch, committee: number[] | Uint32Array) => {
```
Comment (twoeths) [inline packages/beacon-node/src/chain/opPools/aggregatedAttestationPool.ts]: can we drop `number[]` type? same to below

File: packages/state-transition/src/util/shuffle.ts [hunk 2]
```diff
@@ -70,7 +70,7 @@ function setPositionUint32(value: number, buf: Buffer): void {
 }
 
 // Shuffles or unshuffles, depending on the `dir` (true for shuffling, false for unshuffling
-function innerShuffleList(input: ValidatorIndex[], seed: Bytes32, dir: boolean): void {
+function innerShuffleList(input: Uint32Array | ValidatorIndex[], seed: Bytes32, dir: boolean): void {
```
Comment (matthewkeil) [inline packages/state-transition/src/util/shuffle.ts]: Is the `ValidatorIndex[]` type correct here? Same in the other function definitions.

**Comment by twoeths** [comment]
looks good to me, dropped a minor comment @wemeetagain

File: packages/state-transition/src/util/shuffle.ts (hunk 2 above)
Comment (wemeetagain) [reply packages/state-transition/src/util/shuffle.ts]: Yeah, its intentional, its just a wider type than necessary, allows the tests to continue to pass unchanged. Really the implementation relies on an interface like:
```ts
interface Shuffleable {
  [index: number]: number
  length: number
}
```

**Review by nflaig — APPROVED** [review]
LGTM

**Review by twoeths — APPROVED** [review]

**Review by matthewkeil — APPROVED** [review]
Sweet!!! 🎸

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.17.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.17.0) :tada:

**Pull request merged**